File: src/main/java/dev/klomptech/jbark/parser/Parser.java
package dev.klomptech.jbark.parser;

import java.util.ArrayList;
import java.util.List;

import dev.klomptech.jbark.BarkOptions;
import dev.klomptech.jbark.errors.BarkError;
import dev.klomptech.jbark.lexer.Token;
import dev.klomptech.jbark.lexer.TokenType;
import dev.klomptech.jbark.parser.assign.AssignParser;
import dev.klomptech.jbark.parser.collection.PileActionParser;
import dev.klomptech.jbark.parser.collection.PileParser;
import dev.klomptech.jbark.parser.collection.StashParser;
import dev.klomptech.jbark.parser.collection.TakeParser;
import dev.klomptech.jbark.parser.controlflow.ForEachParser;
import dev.klomptech.jbark.parser.controlflow.FunctionParser;
import dev.klomptech.jbark.parser.controlflow.IfParser;
import dev.klomptech.jbark.parser.controlflow.LoopControlParser;
import dev.klomptech.jbark.parser.controlflow.UntilParser;
import dev.klomptech.jbark.parser.controlflow.WhileParser;
import dev.klomptech.jbark.parser.input.InputParser;
import dev.klomptech.jbark.parser.printing.PrintParser;
import dev.klomptech.jbark.parser.share.ShareParser;

public class Parser {

  private final List<Token> tokens;
  private final BarkOptions options;
  private int current = 0;

  private final PrintParser printParser = new PrintParser(this);
  private final IfParser ifParser = new IfParser(this);
  private final WhileParser whileParser = new WhileParser(this);
  private final UntilParser untilParser = new UntilParser(this);
  private final ForEachParser forEachParser = new ForEachParser(this);
  private final LoopControlParser loopControlParser = new LoopControlParser(this);
  private final FunctionParser functionParser = new FunctionParser(this);
  private final InputParser inputParser = new InputParser(this);
  private final AssignParser assignParser = new AssignParser(this);
  private final ShareParser shareParser = new ShareParser(this);
  private final StashParser stashParser = new StashParser(this);
  private final PileParser pileParser = new PileParser(this);
  private final PileActionParser pileActionParser = new PileActionParser(this);
  private final TakeParser takeParser = new TakeParser(this);

  public Parser(final List<Token> tokens) {
    this(tokens, BarkOptions.defaults());
  }

  public Parser(final List<Token> tokens, final BarkOptions options) {
    this.tokens = tokens;
    this.options = options == null ? BarkOptions.defaults() : options;
  }

  public AstNode.Program parse() throws BarkError {
    List<AstNode> statements = new ArrayList<>();
    while (!isAtEnd()) {
      if (isNewline()) {
        advance();
        continue;
      }
      int line = peek().line();
      AstNode node = parseLine();
      if (node instanceof AstNode.LineGroup group) {
        statements.addAll(group.statements());
      } else if (node != null) {
        statements.add(node);
      } else {
        consumeUntilLineEnd();
        warnIfUnrecognizedLine(line);
      }
    }
    return new AstNode.Program(statements);
  }

  // Main processing methods
  private AstNode parseLine() throws BarkError {
    Token token = peek();
    return switch (token.type()) {
      case EOF, NEWLINE -> null;
      case NUMBER -> parseNumberLine();
      case IDENTIFIER -> parseWordLine();
      default -> null;
    };
  }

  private AstNode parseNumberLine() throws BarkError {
    return assignParser.parseLeadingObjectLine();
  }

  private AstNode parseWordLine() throws BarkError {
    // Example: "when she has 2 toys then"
    if (lineHasWord(Keywords.IF_START_WORDS)) {
      return ifParser.parseIfLine();
    }
    // Example: "while she sniffs less toys than 8 then"
    if (lineHasWord(Keywords.WHILE_KEYWORDS)) {
      return whileParser.parseWhileLine();
    }
    // Example: "until she has 0 toys then"
    if (lineHasWord(Keywords.UNTIL_KEYWORDS)) {
      return untilParser.parseUntilLine();
    }
    // Example: "for each treat from her cookie jar then"
    if (currentLineStartsWithAny(Keywords.FOR_WORDS)) {
      return forEachParser.parseForEachLine();
    }
    // Example: "heel" stops a while/until/for-each loop early (break)
    AstNode loopBreak = loopControlParser.parseBreakLine();
    if (loopBreak != null) {
      return loopBreak;
    }
    // Example: "again" ends this lap and starts the next loop round (continue)
    AstNode loopContinue = loopControlParser.parseContinueLine();
    if (loopContinue != null) {
      return loopContinue;
    }
    // Example: "add expects a b returns a" ... bury
    AstNode trick = functionParser.parseTrickLine();
    if (trick != null) {
      return trick;
    }
    // Example: "escape" stops the program
    AstNode exit = inputParser.parseExitLine();
    if (exit != null) {
      return exit;
    }
    // Example: "sleep" or "wait 2 seconds"
    AstNode wait = inputParser.parseWaitLine();
    if (wait != null) {
      return wait;
    }
    // Example: "the labrador listens"
    AstNode listen = inputParser.parseListenLine();
    if (listen != null) {
      return listen;
    }
    // Example: "she passes a toy to the beagle" (one count, one recipient)
    AstNode pass = shareParser.parsePassLine();
    if (pass != null) {
      return pass;
    }
    // Example: "she shares a toy with the beagle" (many counts, split across many recipients)
    AstNode share = shareParser.parseShareLine();
    if (share != null) {
      return share;
    }
    // Example: "my labrador holds 3 toys"
    AstNode stash = stashParser.parseStashLine();
    if (stash != null) {
      return stash;
    }
    // Example: "she takes the top item from the laundry basket"
    AstNode pilePop = pileActionParser.parseGrabTopLine();
    if (pilePop != null) {
      return pilePop;
    }
    // Example: "she puts "slipper" into the laundry basket"
    AstNode pilePush = pileActionParser.parsePushIntoLine();
    if (pilePush != null) {
      return pilePush;
    }
    // Example: "she wants the first biscuit from her cookie jar"
    AstNode take = takeParser.parseTakeLine();
    if (take != null) {
      return take;
    }
    // Example: "the laundry basket holds "shirt", "sock""
    AstNode pile = pileParser.parsePileLine();
    if (pile != null) {
      return pile;
    }
    // Example: "she misplaces a toy", "she gains 2 treats"
    AstNode adjust = assignParser.parseAdjustLine();
    if (adjust != null) {
      return adjust;
    }
    AstNode hasAssign = assignParser.parseHasLine();
    if (hasAssign != null) {
      return hasAssign;
    }
    // Example: "ball 5"
    AstNode objectCount = assignParser.parseObjectCountLine();
    if (objectCount != null) {
      return objectCount;
    }
    // Example: "labrador name is Bimba"
    AstNode nameAssign = assignParser.parseNameLine();
    if (nameAssign != null) {
      return nameAssign;
    }
    // Example: "bark my labrador is loud" reads the trait and prints true/false (does not set it)
    if (lineHasWord(Keywords.PRINT_KEYWORDS)) {
      return printParser.parsePrintLine();
    }
    // Example: "my labrador is loud" sets the trait (no print verb on the line)
    AstNode trait = assignParser.parseTraitLine();
    if (trait != null) {
      return trait;
    }
    // Example: "He is 5 years old"
    AstNode yearsOld = assignParser.parseYearsOldLine();
    if (yearsOld != null) {
      return yearsOld;
    }
    // Example: "She is 2"
    AstNode isAge = assignParser.parseIsAgeLine();
    if (isAge != null) {
      return isAge;
    }
    // Example: "labrador 3" on its own
    AstNode bareAge = assignParser.parseBareAgeLine();
    if (bareAge != null) {
      return bareAge;
    }
    // Example: "memory is count of the items in the cookie jar"
    AstNode storyConstant = assignParser.parseStoryConstantAssignLine();
    if (storyConstant != null) {
      return storyConstant;
    }
    // Example: "my labrador treats is count of the items in the cookie jar"
    AstNode attributeExpr = assignParser.parseAttributeExpressionAssignLine();
    if (attributeExpr != null) {
      return attributeExpr;
    }
    // Example: "So, my wife has a labrador" registers the breed on the line
    AstNode register = assignParser.parseRegisterLine();
    if (register != null) {
      return register;
    }
    return null;
  }

  // Any identifier on this line matches one of the words after normalise
  public boolean lineHasWord(final List<String> words) {
    for (int offset = 0; !isAtEndOrLineEndOffset(offset); offset++) {
      Token token = peekAt(offset);
      if (token.is(TokenType.IDENTIFIER) && words.contains(normalise(token.value()))) {
        return true;
      }
    }
    return false;
  }

  // Warn if the line produced no statement
  private void warnIfUnrecognizedLine(final int line) {
    if (options.strict()) {
      System.err.println("Line " + line + ": the dogs didn't catch that line (story-only words?).");
    }
  }

  // Helper methods
  public Token peek() {
    return tokens.get(current);
  }

  public Token peekAt(final int offset) {
    int index = current + offset;
    if (index >= tokens.size()) {
      return tokens.get(tokens.size() - 1);
    }
    return tokens.get(index);
  }

  public Token advance() {
    return tokens.get(current++);
  }

  public void advanceBy(final int count) {
    for (int i = 0; i < count; i++) {
      advance();
    }
  }

  public boolean isAtEnd() {
    return peek().is(TokenType.EOF);
  }

  private boolean isNewline() {
    return peek().is(TokenType.NEWLINE);
  }

  public boolean isAtEndOrLineEndOffset(final int offset) {
    Token token = peekAt(offset);
    return token.is(TokenType.NEWLINE) || token.is(TokenType.EOF);
  }

  // How many tokens from peekAt(startOffsetFromCursor) until newline (newline not included)
  public int countTokensAhead(final int startOffsetFromCursor) {
    int count = 0;
    while (!isAtEndOrLineEndOffset(startOffsetFromCursor + count)) {
      count++;
    }
    return count;
  }

  public void skipNewlines() {
    while (!isAtEnd() && isNewline()) {
      advance();
    }
  }

  // First identifier on the current line (after normalise) equals word
  public boolean currentLineStartsWith(final String word) {
    Token token = peek();
    return token.is(TokenType.IDENTIFIER) && word.equals(normalise(token.value()));
  }

  public boolean currentLineStartsWithOtherwise() {
    return currentLineStartsWithAny(Keywords.OTHERWISE_WORDS)
        || currentLineStartsWithAny(Keywords.ELSE_WORDS);
  }

  // Elif branch: line starts with otherwise and later has when/if on the same line
  // Example: "otherwise when she has 1 toy then". A "otherwise" alone is final else, not this.
  public boolean currentLineIsOtherwiseWhen() {
    if (!currentLineStartsWithOtherwise()) {
      return false;
    }
    for (int offset = 1; !isAtEndOrLineEndOffset(offset); offset++) {
      Token token = peekAt(offset);
      if (token.is(TokenType.IDENTIFIER) && Keywords.isIfStart(normalise(token.value()))) {
        return true;
      }
    }
    return false;
  }

  public boolean currentLineIsBury() {
    return currentLineStartsWithAny(Keywords.BURY_KEYWORDS);
  }

  // Stop lines while reading a multiline then-body
  public boolean currentLineIsMultilineIfStop() {
    return currentLineStartsWithOtherwise() || currentLineIsBury();
  }

  public boolean currentLineStartsWithAny(final List<String> words) {
    Token token = peek();
    if (!token.is(TokenType.IDENTIFIER)) {
      return false;
    }
    return words.contains(normalise(token.value()));
  }

  // Read one line the same way the main parse loop does (if, print, take, and so on)
  public AstNode parseStatement() throws BarkError {
    return parseLine();
  }

  // LineGroup only happens when one line has two actions (example: whimper print + take from jar)
  // This method splits that into two separate steps in the list
  public void parseStatementInto(final List<AstNode> steps) throws BarkError {
    AstNode stmt = parseStatement();
    if (stmt instanceof AstNode.LineGroup group) {
      steps.addAll(group.statements());
    } else if (stmt != null) {
      steps.add(stmt);
    } else {
      consumeUntilLineEnd();
    }
  }

  // Eat a bury (or alternatives) line and move past its newline
  public void consumeBuryLine() {
    if (currentLineIsBury()) {
      consumeUntilLineEnd();
    }
    skipNewlines();
  }

  public void consumeUntilLineEnd() {
    while (!isAtEnd()) {
      advance();
      if (isNewline()) {
        return;
      }
    }
  }

  public String normalise(final String value) {
    return value.trim().toLowerCase();
  }

  // Skip identifier tokens the dogs do not hear (story filler).
  public int skipIgnoredIdentifiers(final int from, final int to) {
    int pos = from;
    while (pos < to) {
      Token token = peekAt(pos);
      if (!token.is(TokenType.IDENTIFIER)) {
        break;
      }
      String word = normalise(token.value());
      if (!Keywords.isIgnored(word)) {
        break;
      }
      pos++;
    }
    return pos;
  }

  public PrintParser getPrintParser() {
    return printParser;
  }

  public int cursorIndex() {
    return current;
  }

  public void setCurrent(final int index) {
    current = index;
  }

  public List<Token> allTokens() {
    return tokens;
  }

  // Token index where the current line ends (newline or EOF)
  public int lineEndIndex() {
    for (int i = current; i < tokens.size(); i++) {
      if (tokens.get(i).is(TokenType.NEWLINE)) {
        return i;
      }
    }
    return tokens.size();
  }
}

File: pbark/pbark/parser/parser.py
from __future__ import annotations

import sys

from pbark.errors import BarkError
from pbark.lexer import Token, TokenType
from pbark.options import BarkOptions
from pbark.parser import ast_node as AstNode
from pbark.parser import keywords as Keywords
from pbark.parser.assign.assign_parser import AssignParser
from pbark.parser.collection.pile_action_parser import PileActionParser
from pbark.parser.collection.pile_parser import PileParser
from pbark.parser.collection.stash_parser import StashParser
from pbark.parser.collection.take_parser import TakeParser
from pbark.parser.controlflow.foreach_parser import ForEachParser
from pbark.parser.controlflow.function_parser import FunctionParser
from pbark.parser.controlflow.if_parser import IfParser
from pbark.parser.controlflow.loop_control_parser import LoopControlParser
from pbark.parser.controlflow.until_parser import UntilParser
from pbark.parser.controlflow.while_parser import WhileParser
from pbark.parser.input.input_parser import InputParser
from pbark.parser.printing.print_parser import PrintParser
from pbark.parser.share.share_parser import ShareParser


class Parser:
    def __init__(self, tokens: list[Token], options: BarkOptions | None = None) -> None:
        self._tokens = tokens
        self._options = options if options is not None else BarkOptions.defaults()
        self._current = 0
        self._print_parser = PrintParser(self)
        self._if_parser = IfParser(self)
        self._while_parser = WhileParser(self)
        self._until_parser = UntilParser(self)
        self._for_each_parser = ForEachParser(self)
        self._loop_control_parser = LoopControlParser(self)
        self._function_parser = FunctionParser(self)
        self._input_parser = InputParser(self)
        self._assign_parser = AssignParser(self)
        self._share_parser = ShareParser(self)
        self._stash_parser = StashParser(self)
        self._pile_parser = PileParser(self)
        self._pile_action_parser = PileActionParser(self)
        self._take_parser = TakeParser(self)

    def parse(self) -> AstNode.Program:
        statements = []
        while not self.is_at_end():
            if self._is_newline():
                self.advance()
                continue
            line = self.peek().line
            node = self._parse_line()
            if isinstance(node, AstNode.LineGroup):
                statements.extend(node.statements)
            elif node is not None:
                statements.append(node)
            else:
                self.consume_until_line_end()
                self._warn_if_unrecognized_line(line)
        return AstNode.Program(statements)

    def _parse_line(self):
        token = self.peek()
        if token.is_(TokenType.EOF) or token.is_(TokenType.NEWLINE):
            return None
        if token.is_(TokenType.NUMBER):
            return self._assign_parser.parse_leading_object_line()
        if token.is_(TokenType.IDENTIFIER):
            return self._parse_word_line()
        return None

    def _parse_word_line(self):
        if self.line_has_word(Keywords.IF_START_WORDS):
            return self._if_parser.parse_if_line()
        if self.line_has_word(Keywords.WHILE_KEYWORDS):
            return self._while_parser.parse_while_line()
        if self.line_has_word(Keywords.UNTIL_KEYWORDS):
            return self._until_parser.parse_until_line()
        if self.current_line_starts_with_any(Keywords.FOR_WORDS):
            return self._for_each_parser.parse_for_each_line()
        loop_break = self._loop_control_parser.parse_break_line()
        if loop_break is not None:
            return loop_break
        loop_continue = self._loop_control_parser.parse_continue_line()
        if loop_continue is not None:
            return loop_continue
        trick = self._function_parser.parse_trick_line()
        if trick is not None:
            return trick
        exit_node = self._input_parser.parse_exit_line()
        if exit_node is not None:
            return exit_node
        wait = self._input_parser.parse_wait_line()
        if wait is not None:
            return wait
        listen = self._input_parser.parse_listen_line()
        if listen is not None:
            return listen
        pass_node = self._share_parser.parse_pass_line()
        if pass_node is not None:
            return pass_node
        share = self._share_parser.parse_share_line()
        if share is not None:
            return share
        stash = self._stash_parser.parse_stash_line()
        if stash is not None:
            return stash
        pile_pop = self._pile_action_parser.parse_grab_top_line()
        if pile_pop is not None:
            return pile_pop
        pile_push = self._pile_action_parser.parse_push_into_line()
        if pile_push is not None:
            return pile_push
        take = self._take_parser.parse_take_line()
        if take is not None:
            return take
        pile = self._pile_parser.parse_pile_line()
        if pile is not None:
            return pile
        adjust = self._assign_parser.parse_adjust_line()
        if adjust is not None:
            return adjust
        has_assign = self._assign_parser.parse_has_line()
        if has_assign is not None:
            return has_assign
        object_count = self._assign_parser.parse_object_count_line()
        if object_count is not None:
            return object_count
        name_assign = self._assign_parser.parse_name_line()
        if name_assign is not None:
            return name_assign
        if self.line_has_word(Keywords.PRINT_KEYWORDS):
            return self._print_parser.parse_print_line()
        trait = self._assign_parser.parse_trait_line()
        if trait is not None:
            return trait
        years_old = self._assign_parser.parse_years_old_line()
        if years_old is not None:
            return years_old
        is_age = self._assign_parser.parse_is_age_line()
        if is_age is not None:
            return is_age
        bare_age = self._assign_parser.parse_bare_age_line()
        if bare_age is not None:
            return bare_age
        story_constant = self._assign_parser.parse_story_constant_assign_line()
        if story_constant is not None:
            return story_constant
        attribute_expr = self._assign_parser.parse_attribute_expression_assign_line()
        if attribute_expr is not None:
            return attribute_expr
        register = self._assign_parser.parse_register_line()
        if register is not None:
            return register
        return None

    def line_has_word(self, words: list[str]) -> bool:
        offset = 0
        while not self.is_at_end_or_line_end_offset(offset):
            token = self.peek_at(offset)
            if token.is_(TokenType.IDENTIFIER) and self.normalise(token.value) in words:
                return True
            offset += 1
        return False

    def _warn_if_unrecognized_line(self, line: int) -> None:
        if self._options.strict:
            print(
                f"Line {line}: the dogs didn't catch that line (story-only words?).",
                file=sys.stderr,
            )

    def peek(self) -> Token:
        return self._tokens[self._current]

    def peek_at(self, offset: int) -> Token:
        index = self._current + offset
        if index >= len(self._tokens):
            return self._tokens[-1]
        return self._tokens[index]

    def advance(self) -> Token:
        token = self._tokens[self._current]
        self._current += 1
        return token

    def advance_by(self, count: int) -> None:
        self._current += count

    def is_at_end(self) -> bool:
        return self.peek().is_(TokenType.EOF)

    def _is_newline(self) -> bool:
        return self.peek().is_(TokenType.NEWLINE)

    def is_at_end_or_line_end_offset(self, offset: int) -> bool:
        token = self.peek_at(offset)
        return token.is_(TokenType.NEWLINE) or token.is_(TokenType.EOF)

    def count_tokens_ahead(self, start_offset_from_cursor: int) -> int:
        count = 0
        while not self.is_at_end_or_line_end_offset(start_offset_from_cursor + count):
            count += 1
        return count

    def skip_newlines(self) -> None:
        while not self.is_at_end() and self._is_newline():
            self.advance()

    def current_line_starts_with(self, word: str) -> bool:
        token = self.peek()
        return token.is_(TokenType.IDENTIFIER) and word == self.normalise(token.value)

    def current_line_starts_with_otherwise(self) -> bool:
        return self.current_line_starts_with_any(
            Keywords.OTHERWISE_WORDS
        ) or self.current_line_starts_with_any(Keywords.ELSE_WORDS)

    def current_line_is_otherwise_when(self) -> bool:
        if not self.current_line_starts_with_otherwise():
            return False
        offset = 1
        while not self.is_at_end_or_line_end_offset(offset):
            token = self.peek_at(offset)
            if token.is_(TokenType.IDENTIFIER) and Keywords.is_if_start(
                self.normalise(token.value)
            ):
                return True
            offset += 1
        return False

    def current_line_is_bury(self) -> bool:
        return self.current_line_starts_with_any(Keywords.BURY_KEYWORDS)

    def current_line_is_multiline_if_stop(self) -> bool:
        return self.current_line_starts_with_otherwise() or self.current_line_is_bury()

    def current_line_starts_with_any(self, words: list[str]) -> bool:
        token = self.peek()
        if not token.is_(TokenType.IDENTIFIER):
            return False
        return self.normalise(token.value) in words

    def parse_statement(self):
        return self._parse_line()

    def parse_statement_into(self, steps: list) -> None:
        stmt = self.parse_statement()
        if isinstance(stmt, AstNode.LineGroup):
            steps.extend(stmt.statements)
        elif stmt is not None:
            steps.append(stmt)
        else:
            self.consume_until_line_end()

    def consume_bury_line(self) -> None:
        if self.current_line_is_bury():
            self.consume_until_line_end()
        self.skip_newlines()

    def consume_until_line_end(self) -> None:
        while not self.is_at_end():
            self.advance()
            if self._is_newline():
                return

    def normalise(self, value: str) -> str:
        return value.strip().lower()

    def skip_ignored_identifiers(self, from_: int, to: int) -> int:
        pos = from_
        while pos < to:
            token = self.peek_at(pos)
            if not token.is_(TokenType.IDENTIFIER):
                break
            word = self.normalise(token.value)
            if not Keywords.is_ignored(word):
                break
            pos += 1
        return pos

    def get_print_parser(self) -> PrintParser:
        return self._print_parser

    def cursor_index(self) -> int:
        return self._current

    def set_current(self, index: int) -> None:
        self._current = index

    def all_tokens(self) -> list[Token]:
        return self._tokens

    def line_end_index(self) -> int:
        for i in range(self._current, len(self._tokens)):
            if self._tokens[i].is_(TokenType.NEWLINE):
                return i
        return len(self._tokens)
